Blog App: Add authorization rules

mahdinoori2000 opened the pull request:
### What is added

- Install CanCanCan in the project.
- Add a role column to the users table.
- A user can delete a post if it is theirs or if they have an admin role (column role has value "admin"). 
- Add the "Delete" button to the view and make sure that only authorized users can see it
- A user can delete a comment if it is theirs or if they have an admin role (column role has value "admin").
- Add the "Delete" button to the view and make sure that only authorized users can see it.

mahdinoori2000 committed:
Add cancancan gem

mahdinoori2000 committed:
Add role to the application controller

tsheporamantso committed:
Update README file with author contact details

tsheporamantso committed:
Define destroy action for post controller

tsheporamantso committed:
Add conditional statement on ability.rb file

MohammadYaser committed:
Add contact details in readme file

MohammadYaser committed:
check if user can destroy the post

MohammadYaser committed:
Add role field in sign up page

mahdinoori2000 committed:
Add authorization for comments

shahadat3669 commented on app/controllers/posts_controller.rb:
```diff
@@ -1,4 +1,5 @@
 class PostsController < ApplicationController
+  load_and_authorize_resource
   def index
```
- [ ] [Optional] 👏 Good work so far! However, there are some inconsistencies in post counts on the following pages:

- **`users#index`:**
  - The post count for users appears to be incorrect. Please review the code and ensure that the post count accurately reflects the number of posts associated with each user.

- **`user#show`:**
  - Similar to the `users#index` issue, the post count for individual users on the `user#show` page seems inaccurate. Double-check the logic to ensure correct post counts.

- **`post#index`:**
  - The post count displayed on the `post#index` page also appears to be inconsistent. Revisit the code responsible for calculating and displaying post counts to address this discrepancy.


https://github.com/mahdinoori2000/blog-posts/assets/55840999/361cbb98-0c82-43ea-89f0-ff8acb77a527



By addressing these issues, you'll ensure that the post counts on these pages accurately represent the associated posts. Keep up the good work! 🚀🔍

shahadat3669 commented on app/controllers/comments_controller.rb:
```diff
@@ -15,6 +15,10 @@ def create
     end
   end
 
```
- [ ] [Optional]👏 Good progress! Unfortunately, an issue arises when attempting to create a new comment from the UI. The error indicates a `NoMethodError` in `CommentsController#new` with the message "undefined method `posts' for nil:NilClass." Please review your code to identify and resolve the cause of this error. Ensure that the necessary associations or variables are properly initialized and that the `posts` method is accessible within the `CommentsController#new` action. However, I was able to add a comment from the rails console.
![image](https://github.com/mahdinoori2000/blog-posts/assets/55840999/44567b91-c33e-46cb-adf7-d1cfefa30817)

Once addressed, it should enable a smooth experience for creating new comments. Keep up the good work! 🚀🛠️

Gambit142 replied:
- [OPTIONAL] It would be nice to fix this issue because we can't create a comment from the UI or by inputting the URL in the browser pane. The only way to create comments in your application is through seeding in the database. 💻 

You are missing the post and user constraints. 👍 

#### Hint: In the `Comments` controller, you can do this:
```ruby
  def new
    @post = Post.find(params[:post_id])
    @current_user = current_user
  end
```

shahadat3669 commented on app/controllers/comments_controller.rb:
```diff
@@ -15,6 +15,10 @@ def create
     end
   end
 
+  def destroy
+    @comment.destroy
+  end
```
- [ ] 👍 Good progress on your work! However, there's a critical issue that needs attention. When attempting to delete a comment for both admin and author, the application throws an ActiveRecord::InvalidForeignKey error. This error often indicates that there's a foreign key constraint violation, preventing the deletion operation. 

https://github.com/mahdinoori2000/blog-posts/assets/55840999/23666a46-2503-4b39-8ed7-fd6af9a11ea9



Addressing this issue will enhance the overall functionality of your application. Keep up the good work, and feel free to reach out if you have any questions or need further assistance! 🚀🔍

Gambit142 replied:
- [ ] This issue persists and has not been resolved. You need to define the variables properly in the `destroy` action of the `Comments` controller for you to successfully delete a comment. 👍 😎 

#### Hint:

```ruby
  def destroy
    @comment = Comment.find(params[:id])
    @comment.destroy
    redirect_to user_post_path(@comment.post.author, @comment.post)
  end
```

shahadat3669 commented on app/controllers/posts_controller.rb:
```diff
@@ -26,6 +27,11 @@ def create
     end
   end
 
+  def destroy
+    @post.destroy
+    redirect_to user_path(User.find(params[:user_id]))
+  end
```
👍 Good progress on your project! However, there's an issue that needs attention. Currently, if a post has at least one like or comment, it's not possible to delete the post from both user and admin roles. This could be due to constraints or conditions in your code preventing the deletion.

Here are some steps to address this issue:

- [ ] **Check Constraints:** Review your code to identify any constraints or conditions that might be preventing the deletion of a post with likes or comments. Look for conditions in your controller or service that might be restricting the deletion based on associations.

- [ ] **Handle Deletions Appropriately:** Ensure that your deletion logic is handling associated records (likes, comments, etc.) appropriately before attempting to delete the post. You may need to destroy or nullify associated records first.

![image](https://github.com/mahdinoori2000/blog-posts/assets/55840999/3114bb57-471d-47bf-9607-61ca22ee6813)

![image](https://github.com/mahdinoori2000/blog-posts/assets/55840999/73da399d-9cb2-4376-9ee1-3a5ae04b25df)

By addressing this issue, you'll enhance the overall usability and robustness of your application. Keep up the good work, and feel free to reach out if you have any questions or need further assistance! 🚀🛠️

Gambit142 replied:
- [ ] This issue has not been addressed. When I try to delete a `Post` instance that has either likes or comments, the error still shows:

![image](https://github.com/mahdinoori2000/blog-posts/assets/77106962/290a87ff-84fb-4933-9e4b-af387da66012)

---

To fix this, I would advise you to delete the posts and their dependents. You can define this in the `app/models/post.rb` file. Something like this:

```ruby
has_many :likes, dependent: :destroy
has_many :comments, dependent: :destroy
```

shahadat3669 reviewed (changes requested):
Hi @mahdinoori2000, @tsheporamantso, @MohammadYaser, 

Great job so far! 👏 You've done well, and you're almost there for the next project ⏭️, but there are a few issues 🐛 that you still need to work on.

<img  src="https://media2.giphy.com/media/m3864rBwwBTKMAbpn3/giphy.gif" />


## Don't Worry 🙏

It's okay to have bugs in your code. 🆗 Please see this code review as an opportunity to learn practical coding skills and discover new tips that can improve your code quality 💯 and help you **get even better** 💼.

**I will guide you on how to address these areas down below 👇.**

## Highlights
✅ everyone has contributed 
✅ Has **CanCan** Gem in **Gemfile**
✅ Has CanCanCan `ability` class and added role for the user



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadat3669">
        <img src="https://img.shields.io/badge/github-%23121011.svg?style=for-the-badge&logo=github&logoColor=white" alt="GitHub">
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        <img src="https://img.shields.io/badge/linkedin-%230077B5.svg?style=for-the-badge&logo=linkedin&logoColor=white" alt="LinkedIn">
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        <img src="https://img.shields.io/badge/Slack-4A154B?style=for-the-badge&logo=slack&logoColor=white" alt="Slack">
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        <img src="https://img.shields.io/badge/Gmail-D14836?style=for-the-badge&logo=gmail&logoColor=white" alt="Email">
      </a>
    </td>
  </tr>
</table>

## ⚠️ WARNING ⚠️

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

mahdinoori2000 committed:
Add required changes

Gambit142 commented on app/views/posts/show.html.erb:
```diff
@@ -8,13 +8,21 @@
   <% @post.comments.each do |comment| %>
     <div class="card-body">
       <p class="card-text"><%= comment.user.name %>: <%= comment.text %></p>
+      <% if (can? :destroy, comment) %>
+      <%= button_to "DELETE Comment", user_post_path(@user, @post), method: :delete, class: "btn btn-danger" %>
+
+      <% end %>
```
- [ ] The **_URL helper_** you are using for deleting the comment is wrong and this means that you won't be able to delete a comment but the post itself when the `DELETE Comment` button is pressed. This will cause unexpected UI interactions in the application. Kindly use the correct **_URL helper_**. 👍 😎

Gambit142 reviewed (changes requested):
Hi @mahdinoori2000 @tsheporamants @MohammadYaser 👋, 

Great work on making the changes requested by a previous reviewer 👏 

## Highlights
- Setup linters correctly ✔️ 
- Installed `cancancan` gem ✔️ 
- Created Ability class ✔️ 
- Created a role column for `Users` table ✔️ 

You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were simply missed by the previous reviewer and addressing them will improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### Please tag me in your question so I can receive the notification. You can tag me by typing @Gambit142

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

mahdinoori2000 committed:
Add functionlity to redicret the page after creating comment

mahdinoori2000 commented:
Dear @Gambit142 ,

I hope this message finds you well. I appreciate your thorough review and have carefully addressed the issues raised. 
- Regarding creating comments in the UI, I can confirm that users can create comments successfully.

- Regarding the concern about deleting posts with likes or comments, I re-evaluated the functionality and found that users can indeed delete posts even if the post has likes or comments or both.

I have double-checked the implementation, and all required changes from the previous reviewer have been applied. 
Thank you for your time and feedback. 
Please let me know if there are additional adjustments needed.

LionRouge1 commented on config/routes.rb:
```diff
@@ -4,8 +4,8 @@
  
   resources :users, only: [:index, :show] do
-    resources :posts, only: [:index, :show, :new, :create] do
-      resources :comments, only: [:new, :create]
+    resources :posts, only: [:index, :show, :new, :create, :destroy] do
+      resources :comments, only: [:new, :create, :destroy]
     end
   end
```
[OPTIONAL]
- [ ] It would be nice to add the likes routes to the `config/routes.rb` file to allow a user to like a post using the UI.

```rb
  resources :users, only: [:index, :show] do
    resources :posts, only: [:index, :show, :new, :create, :destroy] do
      resources :comments, only: [:new, :create, :destroy]
      resources :likes, only: [:create]
    end
  end
```

LionRouge1 reviewed (approved):
Hi **Team**, 

# STATUS: APPROVED! 🥇 💯 
Great work on making the changes 👏 
You've done well implementing the requested changes

![done](https://media.giphy.com/media/3o7abKhOpu0NwenH3O/giphy.gif)

Your project is complete! There is nothing else to say other than... it's time to merge it :shipit:

## Highlights
- Great job, your project is error free :+1: 
- Nice work you meet all the requirements :1st_place_medal: 
- All linters passed :heavy_check_mark: 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

souad988 reviewed (commented):
Hi @mahdinoori2000 , 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, where a commits from the third member are missing would you please share the answer from success team you need to tell them that the other member is not contributing .

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

mahdinoori2000 commented:
Hi @souad988  , 

While you made a great effort in adding the review, unfortunately, You added the review to a project where it's been approved before.

## Invalid Code Review
